src/api/trace/span_context.rs

```diff
@@ -111,6 +113,116 @@ impl SpanId {
     }
 }
 
+/// TraceState carries system-specific configuration data, represented as a list
+/// of key-value pairs. TraceState allows multiple tracing systems to
+/// participate in the same trace.
+///
+/// Please review the [W3C specification] for details on this field.
+///
+/// [W3C specification]: https://www.w3.org/TR/trace-context/#tracestate-header
+#[cfg_attr(feature = "serialize", derive(Deserialize, Serialize))]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct TraceState {
+    ordered_data: VecDeque<(String, String)>,
+}
+
+impl TraceState {
+    /// Creates a new `TraceState` from the given key-value collection.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use opentelemetry::api;
+    ///
+    /// let kvs = vec![("foo", "bar"), ("apple", "banana")];
+    /// let trace_state: api::TraceState = api::TraceState::from_key_value(kvs);
+    ///
+    /// assert_eq!(trace_state.header(), String::from("foo=bar,apple=banana"))
+    /// ```
+    pub fn from_key_value<T, K, V>(trace_state: T) -> Self
+    where
+        T: IntoIterator<Item = (K, V)>,
+        K: ToString,
+        V: ToString,
+    {
+        TraceState {
+            ordered_data: trace_state
+                .into_iter()
+                .map(|(key, value)| (key.to_string(), value.to_string()))
+                .collect(),
+        }
+    }
+
+    /// Retrieves a value for a given key from the `TraceState` if it exists.
+    pub fn get(&self, key: String) -> Option<&str> {
+        self.ordered_data.iter().find_map(|item| {
+            if item.0.as_str() == key.as_str() {
+                Some(item.1.as_str())
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Inserts the given key-value pair into the `TraceState`. If a value already exists for the
+    /// given key, this updates the value and updates the value's position. Returns the updated
+    /// `TraceState`.
+    pub fn insert(self, key: String, value: String) -> TraceState {
+        let mut trace_state = self.delete(key.clone());
```

Comment:
I just re-read the spec and it looks like:

> All mutating operations MUST return a new TraceState with the modifications applied. 

So we'll have to clone the trace state instead of modifying in place looks like.

Also 

> If invalid value is passed the operation MUST NOT return TraceState containing invalid data

So looks like these methods have to return `Result` types and do some validating

Reply:
So `insert` and `delete` are `FnOnce` functions, I can have them return a `Result<TraceState, ()>`, but that would mean on error the original valid `TraceState` has been moved/lost. I could do `Result<TraceState, TraceState>` which would return a clone of the unmodified input which I guess is better than silently ignoring an operation with an invalid key/value?

Reply:
I think they have to be `Fn` functions and take `&self` to be spec compliant, unless I'm misunderstanding the `All mutating operations MUST return a new TraceState with the modifications applied.` part.

Reply:
Seems like trace state is meant to be immutable so you have copies with the changes instead of mutating the underlying values